src/service_node/http_server.cpp

```diff
@@ -0,0 +1,190 @@
+  }
+
+  storage_server::storage_server(const std::string& db_path)
+    : m_storage(new Storage(db_path))
+  {}
+
+  bool storage_server::handle_http_request_map(const http_request_info& query_info,
+                                               http_response_info& response_info,
+                                               connection_context& m_conn_context)
+  {
+    bool handled = false;
+    // TODO: check if we actually belong to the recipient's swarm
+    switch (query_info.m_http_method) {
+      case epee::net_utils::http::http_method_get:
+        if (query_info.m_URI == "/retrieve") {
+          handled = on_retrieve_message(query_info, response_info);
+        }
+        break;
+
+      case epee::net_utils::http::http_method_post:
+        if (query_info.m_URI == "/store") {
+          handled = on_store_message(query_info, response_info);
+        }
+        break;
+
+      default:
+        break;
+    }
+    return handled;
+  }
+  bool storage_server::handle_http_request(const http_request_info& query_info,
+                                           http_response_info& response,
+                                           connection_context& m_conn_context)
+  {
+    LOG_PRINT_L2("HTTP [" << m_conn_context.m_remote_address.host_str() << "] " << query_info.m_http_method_str << " "
+                          << query_info.m_URI);
+    response.m_response_code = 200;
+    response.m_response_comment = "Ok";
+    if (!handle_http_request_map(query_info, response, m_conn_context)) {
+        response.m_response_code = 404;
+        response.m_response_comment = "Not found";
+    }
+    return true;
+  }
+
+  bool parseHeaders(const http_header_info& header_info,
+                           std::unordered_map<std::string, std::string&> required_fields,
+                           http_response_info& response)
+  {
+    for (const auto& pair : header_info.m_etc_fields) {
+    const auto it = required_fields.find(pair.first);
+      if (it != required_fields.end()) {
+        it->second.assign(pair.second);
+      }
+    }
+
+    std::string missingFields = std::accumulate(
+      std::begin(required_fields),
+      std::end(required_fields),
+      std::string(""),
+      [](std::string result, const std::pair<std::string, std::string>& pair) {
+        if (pair.second.empty())
+          return result += pair.first + ",";
+        return result;
+      });
+
+    if (!missingFields.empty()) {
+      missingFields.pop_back(); // trailing comma
+      response.m_body = "Missing fields in header: " + missingFields;
+      response.m_response_code = 400;
+      return false;
+    }
+
+    return true;
+  }
+
+  bool storage_server::on_store_message(const http_request_info& query_info, http_response_info& response)
+  {
+    std::string recipient;
+    std::string timestamp;
+    std::string powNonce;
+    std::string ttl;
+    std::string messageHash;
+
+    const std::unordered_map<std::string, std::string&> required_fields {
```

Comment:
I would consider using `std::map` instead `std::unordered_map` for trivial cases like this (it requires significantly less heap memory). A vector of pairs might be even better.

Reply:
I didn't know about `std:unordered_map` taking more heap memory than `std::map`, I just wrongly assumed that it would be faster not to sort the elements, without any drawbacks. Thanks for the tip!
I won't use a vector of pairs, thought, because `map::find()` is nicer than save a few nano seconds :)